service_container/service_subscribers.rst

```diff
@@ -0,0 +1,403 @@
+    {
+        protected $locator;
+
+        public function __construct(ContainerInterface $locator)
+        {
+            $this->locator = $locator;
+        }
+
+        public static function getSubscribedServices()
+        {
+            return [
+                'doctrine' => '?'.ManagerRegistry::class,
+                'mailer' => \Swift_Mailer::class,
+                'twig' => Environment::class
+            ];
+        }
+
+
+        abstract public function update($entity);
+
+        /**
+         * Renders a view with Twig
+         */
+        public function render($template, $parameters)
+        {
+            return $this->locator->get('twig')->render($template, $parameters);
+        }
+
+        /**
+         * Sends an email with Swiftmailer
+         */
+        public function send($recipient, $title, $body)
+        {
+            $message = new \Swift_Message();
+
+            // ...
+
+            $this->locator->get('mailer')->send($message);
+        }
+    }
+
+
+With this newly created service subscriber, the updaters can be adapted to
+coincide with the service locator::
+
+    // src/AppBundle/Updates/FooUpdater.php
+    namespace AppBundle\Updates;
+
+    class FooUpdater extends AbstractUpdater
+    {
+        public function update($entity)
+        {
+            // ...
+
+            $body = $this->render('Emails/foo_update.html.twig', [/* ... */]);
+            $this->send($entity->getRecipient(), 'Foo update', $body);
+        }
+    }
+
+    // src/AppBundle/Updates/BarUpdater.php
+    namespace AppBundle\Updates;
+
+    use AppBundle\BarManager;
+    use Psr\Container\ContainerInterface;
+
+    class BarUpdater extends AbstractUpdater
+    {
+        private $barManager;
+
+        public function __construct(ContainerInterface $locator, BarManager $barManager)
+        {
+            parent::__construct($locator);
+
+            $this->barManager = $barManager;
+        }
+
+        public function update($entity)
+        {
+            // ...
+
+            $body = $this->render('Emails/bar_update.html.twig', [/* ... */]);
+            $this->send($entity->getRecipient(), 'Bar update', $body);
+        }
+    }
+
+Optionally, you'll need to add the ``container.service_subscriber`` tag to
+the services definitions of the updaters to enable the service subscribers.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        // app/config/services.yml
+        services:
+            AppBundle\Updates\:
+                resource: '../src/Updates'
+                tags: ['container.service_subscriber']
+
+        // or
+
+        services:
+            AppBundle\Updates\FooUpdater:
+                tags: ['container.service_subscriber']
+
+            AppBundle\Updates\BarUpdater:
+                tags: ['container.service_subscriber']
+                arguments:
+                    - '@Psr\Container\ContainerInterface'
+                    - '@AppBundle\BarManager'
```

Comment:
No need to show any of this - the user will not need to wire this stuff

Reply:
The reason I chose to show this is because it was unclear to me at first how to manually pass the service locator to the the service, but I guess this can be part of the note you suggested in the previous comment so +1.